### PR TITLE
feat: add topologyspreadconstraints for service

### PR DESCRIPTION
**Pull request by gshilei**
## What this PR does / why we need it:
PR for issue https://github.com/KusionStack/kusion/issues/1303
Please help to have a review. Thanks.

### Does this PR introduce a user-facing change?

```release-note

```

### Additional documentation e.g., design docs, usage docs, etc.:

```docs

```

### CODE REVIEW

File: modules/service/common.k [hunk 1]
```diff
@@ -30,6 +31,9 @@ schema WorkloadBase(wl.Workload):
     # The number of containers that should be ran.
     replicas?:                   int
 
+    # TopologySpreadConstraint describes how a group of pods ought to spread across topology domains.
+    topologySpreadConstraints:  {str:tp.TopologySpreadConstraint}
```
Comment (Yangyang96) [inline modules/service/common.k:35]: Is this field a must? If not, need to add a question mark.

Comment (gshilei) [reply modules/service/common.k]: fix.

File: modules/service/topologyspreadconstraint/topology_spread_constraint.k [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+    # NodeTaintsPolicy indicates how we will treat node taints when calculating pod topology spread skew.
+    nodeTaintsPolicy?:            str
+
+    check:
+        maxSkew > 0, "maxSkew must be greater than 0"
+        whenUnsatisfiable in ["DoNotSchedule", "ScheduleAnyway"], "whenUnsatisfiable value is invalid"
+        nodeAffinityPolicy in ["Ignore", "Honor"] if nodeAffinityPolicy, "nodeAffinityPolicy value is invalid"
+        nodeTaintsPolicy in ["Ignore", "Honor"] if nodeTaintsPolicy, "nodeTaintsPolicy value is invalid"
+
```
Comment (Yangyang96) [inline modules/service/topologyspreadconstraint/topology_spread_constraint.k:107]:
```suggestion
minDomains > 0 if minDomains != Undefined, "minDomains must be greater than 0"
```

File: modules/service/topologyspreadconstraint/topology_spread_constraint.k [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    maxSkew:                      int
+
+    # TopologyKey is the key of node labels.
+    topologyKey:                  str
+
+    # WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy the spread constraint.
+    whenUnsatisfiable:            str
+
+    # LabelSelector is used to find matching pods.
+    labelSelector?:               LabelSelector
```
Comment (Yangyang96) [inline modules/service/topologyspreadconstraint/topology_spread_constraint.k:88]: Is `labelSelector` optional? It seems `matchLabelKeys` relies on this field described in K8s doc.
To determine which pods should be considered for topology spreading, a `labelSelector` or both `labelSelector` and `matchLabelKeys` must be specified.

**Comment by Yangyang96** [comment]
The comments are clear and straightforward, and the code is properly formatted! Thank you for your contribution! @gshilei

File: modules/service/topologyspreadconstraint/topology_spread_constraint.k (hunk 3 above)
Comment (gshilei) [reply modules/service/topologyspreadconstraint/topology_spread_constraint.k:88]: I  think labelSelector and  matchLabelKeys are no mandatory binding relationship. K8s define labelSelector and matchLabelKeys are optional, so i follow k8s the definition.

image

File: modules/service/topologyspreadconstraint/topology_spread_constraint.k (hunk 2 above)
Comment (gshilei) [reply modules/service/topologyspreadconstraint/topology_spread_constraint.k:108]: fix

File: modules/service/topologyspreadconstraint/topology_spread_constraint.k (hunk 3 above)
Comment (gshilei) [reply modules/service/topologyspreadconstraint/topology_spread_constraint.k:88]: Add matchLabelKeys check if labelSelector exist.

**Review by liu-hm19 — APPROVED** [review]
lgtm

**Pull request merged**